backends/ldap.go

```diff
@@ -0,0 +1,343 @@
+package backends
+
+import (
+	"fmt"
+	"github.com/go-ldap/ldap/v3"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	"strconv"
+	"strings"
+)
+
+type LDAPClientFactory func(LDAP) (LDAPClient, error)
+
+type LDAPClient interface {
+	Close() error
+	Bind(username, password string) error
+	Search(request *ldap.SearchRequest) (*ldap.SearchResult, error)
+}
+
+type LDAP struct {
+	LDAPClientFactory        LDAPClientFactory
+	LDAPClient               LDAPClient
+	Url                      string
+	BaseDN                   string
+	GroupBaseDN              string
+	BindDN                   string
+	BindPass                 string
+	UserFilter               string
+	GroupFilter              string
+	SuperuserFilter          string
+	AclTopicPatternAttribute string
+	AclAccAttribute          string
+}
+
+func NewLDAP(authOpts map[string]string, logLevel log.Level) (LDAP, error) {
+
+	l, err := NewLDAPWithFactory(authOpts, logLevel, func(l LDAP) (LDAPClient, error) {
+		ldapClient, err := ldap.DialURL(l.Url)
+		return ldapClient, err
+	})
+
+	return l, err
+}
+
+func NewLDAPWithFactory(authOpts map[string]string, logLevel log.Level, ldapClientFactory LDAPClientFactory) (LDAP, error) {
```

Comment:
I understand the desire of decoupling the configuration of the backend with the actual connection to it given your tests mocks them, and I'm not a priori against it, but I must ask:
- Are you coming from Java Factory world (just kidding)?
- Isn't there a way to test the backend against a dev ldap server like ldap-utils or something?

I know it's a hassle to setup a real ldap service, but Postgres' failures have shown that if you don't test against the real thing, you might be fooling yourself because the mock works, but the actual service doesn't.

Reply:
Well, actually I have quite a few java projects XD

I didn't find an easy ldap server to do this with yet. Many ldap are a real hassle to set up in tests. Is ldap-utils a library? I cannot find it. Your point is valid though.

Reply:
Allthough, the more I think about it, the LDAP Factory would bring the possibility to do connection pooling in the future.

Reply:
I think I found a few ldap implementations that would be easy to integrate into the test setup

Reply:
https://github.com/jimlambrt/gldap/tree/main/testdirectory looks pretty nice. I will update the tests

Reply:
Cool, thanks!

Reply:
Well it turns out that testdirectory doesn't support stuff like memberOf and a few other things. I've decided to use proper openldap, which was really painful to configure but I its finally done!!